### PR TITLE
Fix issue 7828: enabling the binary logger adds console output

### LINKED ISSUE #7828: Enabling the binary logger adds console output
### Issue Description

Requesting a binary log (`/bl`) will print a line to the terminal which isn't printed otherwise (when doing quiet builds).

### Steps to Reproduce

```shell
$ msbuild test.csproj /bl /verbosity:quiet /nologo
/Library/Frameworks/Mono.framework/Versions/6.12.0/lib/mono/msbuild/15.0/bin/MSBuild.dll /nologo /bl /verbosity:quiet test.csproj
```

### Expected Behavior

Quiet builds should be quiet:

```shell
$ msbuild test.csproj /verbosity:quiet /nologo
[no output]
```

### Actual Behavior

An additional line of text (seems like the parsed MSBuild command line) is printed.

### Analysis

I think the following two code blocks work together to cause this:

https://github.com/dotnet/msbuild/blob/f25c1ee15e92d7d71074df0dc6ab89ed40442545/src/MSBuild/XMake.cs#L3184-L3187

https://github.com/dotnet/msbuild/blob/f25c1ee15e92d7d71074df0dc6ab89ed40442545/src/MSBuild/XMake.cs#L2344-L2348

The extra line of text is also printed when not enabling the binary logger, but using diagnostic verbosity (but none of the other verbosity levels).

### Versions & Configurations

```
$ msbuild --version
Microsoft (R) Build Engine version 16.10.1 for Mono
Copyright (C) Microsoft Corporation. All rights reserved.

16.10.1.31401%
```

This also happens with `dotnet build`, so here's that version too:

```
$ dotnet --version
6.0.301
```

### PR DESCRIPTION
**Pull request by fanhaipeng**
Fixes #7828 

### Context
Binary logger option overrides verbosity by diagnostic and prints a line even verbosity is set to quiet.

### Changes Made
The fix introduces a variable to store the original verbosity, if the "original verbosity" is quiet, it suppresses the console output when binary logger is requested.

### CODE REVIEW

**Review by rainersigwald — COMMENTED** [review]
Can you please add a regression test, using `RunnerUtilities.ExecMSBuild` like

https://github.com/dotnet/msbuild/blob/cc3db358d34ad4cd1ec0c67e17582d7ca2a15040/src/Build.UnitTests/BinaryLogger_Tests.cs#L168

that validates the stdout output?

File: src/MSBuild/XMake.cs [hunk 1]
```diff
@@ -3181,18 +3186,19 @@ private static ILogger[] ProcessLoggingSwitches
...
             out LoggerVerbosity verbosity,
+            out LoggerVerbosity originalVerbosity,
             int cpuCount,
             out ProfilerLogger profilerLogger,
             out bool enableProfiler
         )
         {
             // if verbosity level is not specified, use the default
-            verbosity = LoggerVerbosity.Normal;
+            originalVerbosity = verbosity = LoggerVerbosity.Normal;
```
Comment (Forgind) [inline src/MSBuild/XMake.cs:3196]: nit:
I'd prefer for these to be on separate lines like:
verbosity = LoggerVerbosity.Normal;
originalVerbosity = verbosity;

File: src/MSBuild/XMake.cs [hunk 2]
```diff
@@ -2428,7 +2433,7 @@ out enableProfiler
                         Console.WriteLine(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("PickedUpSwitchesFromAutoResponse", autoResponseFileName));
                     }
 
-                    if (verbosity == LoggerVerbosity.Diagnostic)
+                    if (verbosity == LoggerVerbosity.Diagnostic && originalVerbosity != LoggerVerbosity.Quiet)
```
Comment (Forgind) [inline src/MSBuild/XMake.cs:2436]: Maybe just:
```suggestion
                    if (originalVerbosity == LoggerVerbosity.Diagnostic)
```

?

**Review by Forgind — APPROVED** [review]
LGTM!

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7989) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7989) before we can accept your contribution.<br/><br/>:x: fanhaipeng [sign now](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7989)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/msbuild?pullRequest=7989) it.</sub>

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7989) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7989) before we can accept your contribution.<br/><br/>:x: fanhaipeng [sign now](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7989)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/msbuild?pullRequest=7989) it.</sub>

File: src/MSBuild/XMake.cs (hunk 1 above)
Comment (fanhaipeng) [reply src/MSBuild/XMake.cs]: ### thanks for the comment, I've resolved in the latest commit

File: src/MSBuild/XMake.cs (hunk 2 above)
Comment (fanhaipeng) [reply src/MSBuild/XMake.cs:2436]: when bl is set and other verbosity than quiet/diagnostic is set, the output is still wanted.

**Comment by fanhaipeng** [comment]
Hi @Forgind / @rainersigwald, the failed test case is from MultiplePlugins, should not be related to my change, what next step do you recommend? Thanks

_**Microsoft.Build.Engine.UnitTests.ProjectCache.ProjectCacheTests.MultiplePlugins
System.IndexOutOfRangeException : Index was outside the bounds of the array.**_

**Comment by Forgind** [comment]
@fanhaipeng, I reran it, and it passed.

/cc: @dfederm for awareness on the flaky test.

File: src/MSBuild/XMake.cs (hunk 2 above)
Comment (rainersigwald) [reply src/MSBuild/XMake.cs:2436]: This is interesting. My instinct was to suppress the message for "quiet", "minimal" and "normal" but leave it for "detailed" and "diagnostic". Why "quiet" only?

Comment (fanhaipeng) [reply src/MSBuild/XMake.cs:2436]: Hmm... we need more requirement clarifications. My understanding of the requirement is solely from the #7828 statement. It seems we three have different understandings, let me know how we can make an agreement. @rainersigwald  / @Forgind

Comment (rainersigwald) [reply src/MSBuild/XMake.cs:2436]: @Forgind and I actually talked about this this morning, and I think it's ok for it to be there for "diagnostic" only.

**Review by Forgind — APPROVED** [review]
LGTM!

**Review by rainersigwald — APPROVED** [review]
I'd like to hold this for 17.5/7.0.200 since it's a fairly visible behavior change, but this looks good, thanks!

**Pull request merged**